### PR TITLE
[Merged by Bors] - feat: asymptotics on product filters

### PR DESCRIPTION
**Pull request by llllvvuu**
Content from PNT. [Motivating example](https://github.com/AlexKontorovich/PrimeNumberTheoremAnd/blob/15f08ef06df37286bf8636b4573e84f69994aa4b/PrimeNumberTheoremAnd/PerronFormula.lean#L385-L503) is asymptotics of an integral along x, as y -> some filter.

---

- [x] depends on: #10285

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Asymptotics/Uniform.lean [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+variable [Norm E] [Norm F] {f : ι × α → E} {g : α → F} {l : Filter α}
+
+/-- If f = O(g) uniformly on `s`, then f_i = O(g) for any `i ∈ s`.` -/
+theorem isBigO_of_isBigOUniformly (h : f =O[𝓟 s ×ˢ l] (g ∘ Prod.snd)) {i : ι} (hi : i ∈ s) :
+    (fun x ↦ f (i, x)) =O[l] g := by
+  obtain ⟨C, hC⟩ := h.bound
+  obtain ⟨t, htl, ht⟩ := hC.exists_mem
+  obtain ⟨u, hu, v, hv, huv⟩ := Filter.mem_prod_iff.mp htl
+  refine isBigO_iff.mpr ⟨C, Filter.eventually_iff_exists_mem.mpr ⟨v, hv, ?_⟩⟩
+  exact fun y hy ↦ ht _ <| huv ⟨hu hi, hy⟩
```
Comment (j-loreaux) [inline Mathlib/Analysis/Asymptotics/Uniform.lean:38]: I guess I don't really understand the restriction to principal filters here. For example, a slight modification yields:
```lean
/-- If `f = O(g)` eventually with respect to `l' ×ˢ s`, then `f_i = O(g)` eventually in `l'`. -/
theorem foo {l' : Filter ι} (h : f =O[l' ×ˢ l] (g ∘ Prod.snd)) :
    ∀ᶠ i in l', (fun x ↦ f (i, x)) =O[l] g := by
  obtain ⟨C, hC⟩ := h.bound
  obtain ⟨t, htl, ht⟩ := hC.exists_mem
  obtain ⟨u, hu, v, hv, huv⟩ := Filter.mem_prod_iff.mp htl
  filter_upwards [hu] with i hi
  refine isBigO_iff.mpr ⟨C, Filter.eventually_iff_exists_mem.mpr ⟨v, hv, ?_⟩⟩
  exact fun y hy ↦ ht _ <| huv ⟨hi, hy⟩
```
from which you can get:
```lean
/-- If f = O(g) uniformly on `s`, then f_i = O(g) for any `i ∈ s`.` -/
theorem isBigO_of_isBigOUniformly (h : f =O[𝓟 s ×ˢ l] (g ∘ Prod.snd)) {i : ι} (hi : i ∈ s) :
    (fun x ↦ f (i, x)) =O[l] g := by
  revert i
  rw [← eventually_principal]
  simpa using foo h
```

File: Mathlib/Analysis/Asymptotics/Asymptotics.lean [hunk 2]
```diff
@@ -1938,12 +1938,18 @@ theorem isLittleO_const_id_cobounded (c : F'') :
   isLittleO_const_left.2 <| .inr tendsto_norm_cobounded_atTop
 #align asymptotics.is_o_const_id_comap_norm_at_top Asymptotics.isLittleO_const_id_cobounded
 
-theorem isLittleO_const_id_atTop (c : E'') : (fun _x : ℝ => c) =o[atTop] id :=
-  isLittleO_const_left.2 <| Or.inr tendsto_abs_atTop_atTop
+theorem isLittleO_const_id_cocompact [ProperSpace F'']
+    (c : E'') : (fun _x : F'' => c) =o[cocompact F''] id :=
+  isLittleO_const_left.2 <| Or.inr tendsto_norm_cocompact_atTop
+
+theorem isLittleO_const_id_atTop [LinearOrder F''] [NoMaxOrder F''] [ClosedIciTopology F'']
```
Comment (j-loreaux) [inline Mathlib/Analysis/Asymptotics/Asymptotics.lean:1945]: Can you name a bunch of these results accorinding to the suggestion [here](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Naming.20convention/near/447491526) making things `protected` as appropriate?
```suggestion
theorem IsLittleO.const_id_atTop [LinearOrder F''] [NoMaxOrder F''] [ClosedIciTopology F'']
```

**Review by j-loreaux — COMMENTED** [review]
I may have more to say later, but I wanted to get a review started since this has been sitting so long.

File: Mathlib/Analysis/Asymptotics/Uniform.lean (hunk 1 above)
Comment (llllvvuu) [reply Mathlib/Analysis/Asymptotics/Uniform.lean]: Yeah, this PR wasn't in the right generality, so I've reworked it.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

File: Mathlib/MeasureTheory/Integral/Asymptotics.lean [hunk 3]
```diff
@@ -49,6 +50,51 @@ theorem _root_.Asymptotics.IsBigO.integrable (hfm : AEStronglyMeasurable f μ)
   rewrite [← integrableAtFilter_top] at *
   exact hf.integrableAtFilter ⟨univ, univ_mem, hfm.restrict⟩ hg
 
+namespace Asymptotics
+
+variable {ι : Type*} [MeasurableSpace ι] {f : ι × α → E} {s : Set ι} {μ : Measure ι}
+
+/-- Let `f : X x Y → Z`. If as y → l, f(x, y) = O(g(y)) uniformly on `s : Set X` of finite measure,
+then f is eventually (as y → l) integrable along `s`. -/
```
Comment (j-loreaux) [inline Mathlib/MeasureTheory/Integral/Asymptotics.lean:58]:
```suggestion
/-- Let `f : X x Y → Z`. If as `y` tends to `l`, `f(x, y) = O(g(y))` uniformly on `s : Set X`
of finite measure, then f is eventually (as `y` tends to `l`) integrable along `s`. -/
```

File: Mathlib/MeasureTheory/Integral/Asymptotics.lean [hunk 4]
```diff
@@ -49,6 +50,51 @@ theorem _root_.Asymptotics.IsBigO.integrable (hfm : AEStronglyMeasurable f μ)
...
+  haveI : IsFiniteMeasure (μ.restrict s) := ⟨Measure.restrict_apply_univ s ▸ hμ⟩
+  refine Integrable.mono' (integrable_const (C * ‖g x‖)) (hw x hx.1) ?_
+  filter_upwards [MeasureTheory.self_mem_ae_restrict hs]
+  intro y hy
+  exact ht (y, x) <| huv ⟨hu hy, hx.2⟩
+
+variable [NormedSpace ℝ E] [NormedAddCommGroup F] [NormedSpace ℝ F]
+
+/-- Let `f : X x Y → Z`. If as y → l, f(x, y) = O(g(y)) uniformly on `s : Set X` of finite measure,
+then the integral of f along s is O(g(y)). -/
```
Comment (j-loreaux) [inline Mathlib/MeasureTheory/Integral/Asymptotics.lean:78]:
```suggestion
/-- Let `f : X x Y → Z`. If as `y` tends to `l`, `f(x, y) = O(g(y))` uniformly on `s : Set X`
of finite measure, then the integral of `f` along `s` is `O(g(y))`. -/
```

**Review by j-loreaux — COMMENTED** [review]
Since the build failed anyway, do you mind making these docstring changes?

bors d+

**Pull request closed without merging**